File: api/images/image_manager/IHumanPart.py
from typing import Literal, TypedDict, TypeAlias

# from api.gptAI.HumanInformation import CharacterModeState

BodyUnitKey: TypeAlias = str #足、腕、胴、頭、口に順番識別用番号などが付いたもの。例：3_30_耳　4_40_腕など
BodyUnitVariationKey: TypeAlias = str #口(笑顔)、口(驚き)など,"130_ほほえみ"　など
BodyUnitVariationOnOff: TypeAlias = Literal["on", "off"] #口(笑顔)がonかoffか
BodyUnitValue: TypeAlias = dict[BodyUnitVariationKey, BodyUnitVariationOnOff]
PoseInfo: TypeAlias = dict[BodyUnitKey, BodyUnitValue]


BodyUnitVariationFileKey: TypeAlias = str #"30_ﾌﾞﾝﾌﾞﾝ.json"または30_ﾌﾞﾝﾌﾞﾝ.pngなど。ImageNameの先頭の数字を10倍にしたものがこの文字列の先頭の数字になっている。もう使ってないかもしれない
BodyUnitVariationFileType: TypeAlias = Literal["json", "gazou"] #30_ﾌﾞﾝﾌﾞﾝ.json"または30_ﾌﾞﾝﾌﾞﾝ.pngなどがjsonかgazouか
BodyUnitAllFileTypeInfo: TypeAlias = dict[BodyUnitVariationFileKey, BodyUnitVariationFileType]
AllBodyFileInfo: TypeAlias = dict[BodyUnitKey, BodyUnitAllFileTypeInfo]

class BodyFolderFileName(TypedDict):
    folder_name: str
    file_name: str

class PakuStateSpecies(TypedDict):
    開候補: list[BodyFolderFileName]
    閉: list[BodyFolderFileName]

class OnomatopeiaActionSetting(TypedDict):
    パク:PakuStateSpecies
    パチ:PakuStateSpecies
    ぴょこ:PakuStateSpecies

class NowOnomatopoeiaActionSetting(TypedDict):
    パク:list[BodyFolderFileName]
    パチ:list[BodyFolderFileName]
    ぴょこ:list[BodyFolderFileName]

class HumanDataSetting(TypedDict):
    lip_sync: Literal['パクパク']

class InitImageInfo(TypedDict):
    init: PoseInfo
    all_data: AllBodyFileInfo
    OnomatopeiaActionSetting:OnomatopeiaActionSetting
    NowOnomatopoeiaActionSetting:NowOnomatopoeiaActionSetting
    setting: HumanDataSetting|None

ImageBinary: TypeAlias = bytes #画像のバイナリデータ。strかもしれないが要検証
ImageName: TypeAlias = str #画像のファイル名。"3_ﾌﾞﾝﾌﾞﾝ.png"など
class ImageInfo (TypedDict):
    name: ImageName
    x: int
    y: int
    width: int
    height: int
    z_index: int
    initial_display: bool

class BodyUnitVariationImageInfo (TypedDict):
    img: ImageBinary
    json: ImageInfo

BodyUnitVariationImages: TypeAlias = dict[BodyUnitVariationKey, BodyUnitVariationImageInfo]
BodyPartsImages: TypeAlias = dict[BodyUnitKey, BodyUnitVariationImages]

class HumanData(TypedDict):
    body_parts_iamges: BodyPartsImages
    init_image_info: InitImageInfo
    front_name: str
    char_name: str

File: app-ts/src/ValueObject/IHumanPart.ts
import { ExtendedMap } from "../Extend/extend_collections";
import { ICharacterModeState } from "../UiComponent/CharaInfoSelecter/ICharacterInfo";

export type BodyUnitKey = string; // 足、腕、胴、頭、口に順番識別用番号などが付いたもの。例：3_30_耳　4_40_腕など
export type BodyUnitVariationKey = string; // 口(笑顔)、口(驚き)など,"130_ほほえみ"　など
export type BodyUnitVariationOnOff = "on" | "off"; // 口(笑顔)がonかoffか
export type BodyUnitValue = Record<BodyUnitVariationKey, BodyUnitVariationOnOff>;
export type PoseInfoKey = string; // "init"、"ぐるぐる目"、"水着"など
export type PoseInfo = Record<BodyUnitKey, BodyUnitValue>;
export type PoseInfoMap = ExtendedMap<BodyUnitKey, BodyUnitValue>;

export type BodyUnitVariationFileKey = string; // "30_ﾌﾞﾝﾌﾞﾝ.json"または30_ﾌﾞﾝﾌﾞﾝ.pngなど。ImageNameの先頭の数字を10倍にしたものがこの文字列の先頭の数字になっている。もう使ってないかもしれない
export type BodyUnitVariationFileType = "json" | "gazou"; // "30_ﾌﾞﾝﾌﾞﾝ.json"または30_ﾌﾞﾝﾌﾞﾝ.pngなどがjsonかgazouか
export type BodyUnitAllFileTypeInfo = Record<BodyUnitVariationFileKey, BodyUnitVariationFileType>;
export type AllBodyFileInfo = Record<BodyUnitKey, BodyUnitAllFileTypeInfo>;

export interface BodyFolderFileName {
    folder_name: string;
    file_name: string;
}

export interface PakuStateSpecies {
    開候補: BodyFolderFileName[];
    閉: BodyFolderFileName[];
}

export interface OnomatopeiaActionSetting {
    パク: PakuStateSpecies;
    パチ: PakuStateSpecies;
    ぴょこ: PakuStateSpecies;
}

export interface NowOnomatopoeiaActionSetting {
    パク: BodyFolderFileName[];
    パチ: BodyFolderFileName[];
    ぴょこ: BodyFolderFileName[];
}

export interface HumanDataSetting {
    lip_sync: "パクパク"
}

export interface InitImageInfo {
    init: PoseInfo;
    all_data: AllBodyFileInfo;
    OnomatopeiaActionSetting:OnomatopeiaActionSetting;
    NowOnomatopoeiaActionSetting:NowOnomatopoeiaActionSetting;
    setting: HumanDataSetting|null;
}

export type ImageBinary = Uint8Array; // 画像のバイナリデータ。strかもしれないが要検証
export type ImageName = string; // 画像のファイル名。"3_ﾌﾞﾝﾌﾞﾝ.png"など

export interface ImageInfo {
    name: ImageName;
    x: number;
    y: number;
    width: number;
    height: number;
    z_index: number;
    initial_display: boolean;
}

export interface BodyUnitVariationImageInfo {
    img: ImageBinary;
    json: ImageInfo;
}

export type BodyUnitVariationImages = Record<BodyUnitVariationKey, BodyUnitVariationImageInfo>;
export type BodyUnitVariationImagesMap = ExtendedMap<BodyUnitVariationKey, BodyUnitVariationImageInfo>;

export function convertBodyUnitVariationImagesToMap(bodyUnitVariationImages: BodyUnitVariationImages): BodyUnitVariationImagesMap {
    return new ExtendedMap(Object.entries(bodyUnitVariationImages).sort(
        (a, b) => {
            const keyA = parseInt(a[0].split('_')[0]);
            const keyB = parseInt(b[0].split('_')[0]);
            console.log(keyA, keyB);
            return keyA - keyB;
        }
    ));
}

export type BodyPartsImages = Record<BodyUnitKey, BodyUnitVariationImages>;

export interface HumanData {
    body_parts_iamges: BodyPartsImages;
    init_image_info: InitImageInfo;
    front_name: string;
    char_name: string;
}

export interface CharaCreateData{
    humanData: HumanData;
    characterModeState: ICharacterModeState;
}

export interface CharaCreateDataResponse {
    succese_mode: "成功" | "名前が無効" | "名前を指定してください";
    message: string;
    charaCreateData: CharaCreateData|null;
}

export interface HumanBodyCanvasCssStylePosAndSize {
    height: string; 
    top: string; 
    left: string;
}